From 09d3b2984a00790902747b22af4a0524e1d1756b Mon Sep 17 00:00:00 2001
From: Stanislaw Kardach <skardach@marvell.com>
Date: Fri, 24 Jan 2020 16:16:38 +0100
Subject: [PATCH 0360/1921] octeontx2-af: fix memory leaks on probe failure

Ensure all allocated memory is freed when given HW block initialization
fails.

Change-Id: I058301b8ed7b1cd827ce0f045e1557163c2073cc
Signed-off-by: Stanislaw Kardach <skardach@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/22296
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/af/rvu.c   | 25 ++++++++++++-------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index 0d7189a8b1c2..4d2a28860a4b 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -943,44 +943,51 @@ static int rvu_setup_hw_resources(struct rvu *rvu)
 
 	err = rvu_npc_init(rvu);
 	if (err)
-		goto fwdata_err;
+		goto npc_err;
 
 	err = rvu_cgx_init(rvu);
 	if (err)
-		goto fwdata_err;
+		goto cgx_err;
 
 	/* Assign MACs for CGX mapped functions */
 	rvu_setup_pfvf_macaddress(rvu);
 
 	err = rvu_npa_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto npa_err;
 
 	err = rvu_nix_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto nix_err;
 
 	err = rvu_sso_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto sso_err;
 
 	err = rvu_tim_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto sso_err;
 
 	err = rvu_cpt_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto sso_err;
 
 	err = rvu_sdp_init(rvu);
 	if (err)
-		goto cgx_err;
+		goto sso_err;
 
 	return 0;
 
+sso_err:
+	rvu_sso_freemem(rvu);
+nix_err:
+	rvu_nix_freemem(rvu);
+npa_err:
+	rvu_npa_freemem(rvu);
 cgx_err:
 	rvu_cgx_exit(rvu);
-fwdata_err:
+npc_err:
+	rvu_npc_freemem(rvu);
 	rvu_fwdata_exit(rvu);
 msix_err:
 	rvu_reset_msix(rvu);
-- 
2.31.1

